apps/proto/lib/lexical/proto/macros/typespec.ex

```diff
@@ -1,7 +1,213 @@
 defmodule Lexical.Proto.Macros.Typespec do
-  def build(_opts \\ []) do
+  def typespec(opts \\ [], env \\ nil)
+
+  def typespec([], _) do
     quote do
-      @type t :: %__MODULE__{}
+      %__MODULE__{}
+    end
+  end
+
+  def typespec(opts, env) when is_list(opts) do
+    typespecs =
+      for {name, type} <- opts,
+          name != :.. do
+        {name, do_typespec(type, env)}
+      end
+
+    quote do
+      %__MODULE__{unquote_splicing(typespecs)}
+    end
+  end
+
+  def typespec(typespec, env) do
+    quote do
+      unquote(do_typespec(typespec, env))
+    end
+  end
+
+  def choice(options, env) do
+    do_typespec({:one_of, [], [options]}, env)
+  end
+
+  def keyword_constructor_options(opts, env) do
+    for {name, type} <- opts,
+        name != :.. do
+      {name, do_typespec(type, env)}
+    end
+    |> or_types()
+  end
+
+  defp do_typespec([], _env) do
+    # This is what's presented to typespec when a response has no results, as in the Shutdown response
+    nil
+  end
+
+  defp do_typespec(nil, _env) do
+    quote(do: nil)
+  end
+
+  defp do_typespec({:boolean, _, _}, _env) do
+    quote(do: boolean())
+  end
+
+  defp do_typespec({:string, _, _}, _env) do
+    quote(do: String.t())
+  end
+
+  defp do_typespec({:integer, _, _}, _env) do
+    quote(do: integer())
+  end
+
+  defp do_typespec({:float, _, _}, _env) do
+    quote(do: float())
+  end
+
+  defp do_typespec({:__MODULE__, _, nil}, env) do
+    env.module
+  end
+
+  defp do_typespec({:optional, _, [optional_type]}, env) do
+    quote do
+      unquote(do_typespec(optional_type, env)) | nil
+    end
+  end
+
+  defp do_typespec({:__aliases__, _, raw_alias} = aliased_module, env) do
+    expanded_alias = Macro.expand(aliased_module, env)
+
+    case List.last(raw_alias) do
+      :Position ->
+        other_alias =
+          case expanded_alias do
+            Lexical.Document.Range ->
+              Lexical.Protocol.Types.Range
+
+            _ ->
+              Lexical.Document.Range
+          end
+
+        quote do
+          unquote(expanded_alias).t() | unquote(other_alias).t()
+        end
+
+      :Range ->
+        other_alias =
+          case expanded_alias do
+            Lexical.Document.Range ->
+              Lexical.Protocol.Types.Range
+
+            _ ->
+              Lexical.Document.Range
+          end
+
+        quote do
+          unquote(expanded_alias).t() | unquote(other_alias).t()
+        end
+
+      _ ->
+        quote do
+          unquote(expanded_alias).t()
+        end
+    end
+  end
+
+  defp do_typespec({:literal, _, value}, _env) when is_atom(value) do
+    quote do
+      unquote(value)
```

Comment:
There are a handful of quote blocks like this one of the form:

```elixir
quote do
  unquote(thing)
end
```

All of these are equivalent to just returning `thing` without any additional quoting/unquoting. If it's a stylistic choice to make it clear that we're returning something that can be used as a quoted form, that's fine!